File: drills/elixir/solutions/nc127_reorder_list.exs
defmodule Solution do
  # Three separate steps, each of which is its own drill: find the middle,
  # reverse the back half, then interleave. That decomposition is the whole
  # trick -- none of the three needs to know about the others.
  def reorder_list(values) do
    half = div(length(values) + 1, 2)
    {front, back} = Enum.split(values, half)
    interleave(front, Enum.reverse(back))
  end

  defp interleave([], rest), do: rest
  defp interleave(rest, []), do: rest
  defp interleave([a | a_rest], [b | b_rest]), do: [a, b | interleave(a_rest, b_rest)]
end

File: drills/go/solutions/nc127_reorder_list.go
package main

func reorderList(head *ListNode) {
	if head == nil || head.Next == nil {
		return
	}
	// Find the middle, reverse the second half, then interleave the two.
	slow, fast := head, head
	for fast.Next != nil && fast.Next.Next != nil {
		slow, fast = slow.Next, fast.Next.Next
	}
	var second *ListNode
	for node := slow.Next; node != nil; {
		node.Next, second, node = second, node, node.Next
	}
	slow.Next = nil
	first := head
	for second != nil {
		first.Next, second.Next, first, second = second, first.Next, first.Next, second.Next
	}
}
